Use PHP-FPM instead of mod_php

oparoz opened the pull request:
Here is a brand new PHP component for the ownCloud snap.
- Uses PHP-FPM and Apache MPM event
- It makes some modifications to the Apache configs in order to be able to work hand in hand with Apache
- It introduces some structure for the location of logs, sockets and pids
- It introduces an auto-configuration script to help deliver an acceptable setup from the get go
- It lets users modify their configs, so that they can optimise them for their work load/case
### Deliveries
- [x] Configure Apache in mpm_event mode #27 
- [x] Remove the mod_php module
- [x] Create a part for building PHP
- [x] Add JPEG and TrueType support in PHP #26
- [x] Optimise PHP build
- [x] Create a PHP-FPM daemon
- [x] Create a phpinfo command
- [x] Automatically generate a default PHP-FPM config, tuned for the hardware the snap is installed on #28
- [x] Automatically generate a default Apache MPM event config, tuned for the hardware the snap is installed on #28
- [x] Build a file structure allowing for users to customise their configuration #41
- [x] Delete `.user.ini` and migrate settings to config files 
### Test plan

Snaps of this PR for your beta PiClouds will be available once Snapcraft 2.8 has been packaged as a PPA.

Sideloading instructions:
https://github.com/owncloud/pi-image/wiki/Sideloading-snaps-for-testing
- Make sure the platform.conf files are properly populated
- After setup, go to the admin section and you should only get warnings about the missing Redis and HTTPS components
- Upload a few PNG, GIF and JPEGs, you should get previews
- Sync a 123MB folder using the sync client. The device should never become unresponsive
### TODO
- [ ] Approve #28
- [ ] Maybe remove more assets such as documentation from PHP
- [x] Fix `if [ $1 -gt 0 ]; then` to make it work when the value is empty
### Examples configs

Pi 2

```
##################################################################
# HARDWARE
##################################################################
Total memory 921
Number of CPU cores 4
CPU speed 900
BogoMIPS 38
##################################################################
# PHP-FPM
##################################################################
The MAXIMUM number of idle PHP-FPM processes 4
The MINIMUM number of idle PHP-FPM processes 2
The number of PHP-FPM child processes created on startup 3
The maximum number of PHP-FPM child processes (max simultaneous requests) 8
The maximum amount of memory allowed PER PHP-FPM PROCESS 131M
##################################################################
# APACHE MPM_EVENT
##################################################################
The maximum number of Apache child processes 12
The number of Apache threads per child process 25
The maximum number of simultaneous requests 300
The MINIMUM number of idle Apache processes 18
The MAXIMUM number of idle Apache processes 60
The number of Apache child processes created on startup 3
```

oparoz commented:
Needing some code review for now @kyrofa 
@LukasReschke @Xenopathic @Gomez for the PHP config

kyrofa commented on snapcraft.yaml:
```diff
@@ -55,59 +75,99 @@ parts:
   boost:
     plugin: copy
-    source: http://sourceforge.net/projects/boost/files/boost/1.59.0/boost_1_59_0.tar.gz
+    source: http://liquidtelecom.dl.sourceforge.net/project/boost/boost/1.59.0/boost_1_59_0.tar.gz
```
Is there a reason for this change?

oparoz replied:
Sorry, leftovers from yesterday when the builders were misbehaving

oparoz commented on src/owncloud/setup_owncloud:
```diff
@@ -1,4 +1,9 @@
+# This script is called by /bin/envvars
+############################################
+#
+. $SNAP/setup/filesystem.sh
```
Ideally, I'd like to be able to place this is a "snap setup script" so that it doesn't need to be added to every daemon.

oparoz replied:
There is no setup script panned for Snappy, so this has to be included in every daemon script

kyrofa commented:
> It lets users modify their configs, so that they can optimise them for their work load/case

I thought we agreed to hold off on this until we agreed on a good system for upgrades? I have two problems with this:
1. We haven't agreed on a good system for upgrades, and
2. It dilutes the purpose of this PR, which is to use php-fpm instead of mod_php. It's trying to tackle multiple issues with a single change. Can we try to keep it 1-to-1 for review, changelog, git-blame, and rollback purposes?

oparoz commented on src/setup/hw_rating.sh:
```diff
@@ -0,0 +1,112 @@
+# Licence GPL 3.0
+##################################################################
+validate_params() {
+	if [ $1 -gt 0 ]; then
```
Needs a better check, added to TODO

oparoz commented:
> We haven't agreed on a good system for upgrades, and

Correct and this PR can wait for this agreement and it's related PR(s). ownCloud doesn't work properly on various hardware configs without the ability to make changes to those config files.

> It dilutes the purpose of this PR, which is to use php-fpm instead of mod_php

I agree, but I needed to put the files somewhere. So if we agree on the filesystem structure and the way to create it, PRs to address the problem can be merged and this will be rebased on top of those changes.
In the mean time, testers have something to test, tune and provide feedback on.

oparoz commented on src/setup/filesystem.sh:
```diff
@@ -0,0 +1,101 @@
+	PHP_START_SERVERS_DEFAULT=2
+	PHP_MAX_CHILDREN_DEFAULT=5
+	PHP_MEMORY_LIMIT_DEFAULT=128
+
```
Setting up defaults to fall back to in case of failure in the autodetect script

oparoz commented on src/setup/filesystem.sh:
```diff
@@ -0,0 +1,101 @@
+
+	# Calculates config values for PHP-FPM and Apache
+	. $SNAP/setup/hw_rating.sh
+	if [ "$?" != 0 ] ; then
```
The autodetect script failed, so we use the default values

oparoz commented on src/setup/hw_rating.sh:
```diff
@@ -0,0 +1,105 @@
+
+# Amount of RAM Apache can use
+APACHE_AVAIL_RAM=$(awk "BEGIN {printf "'"%d"'", $MEM_TOTAL * 0.2 }")
+APACHE_MAX_ALLOWED_RAM=2000
```
We set a limit to the amount of RAM Apache can use.
Currently 2G

oparoz commented on src/setup/hw_rating.sh:
```diff
@@ -0,0 +1,105 @@
+echo "# PHP-FPM"
+echo "##################################################################"
+# Determined in advance via "ps -ylC php-fpm --sort:rss"
+PHP_PROCESS_SIZE=65
```
We have to determine the average size of a PHP process in advance

oparoz commented on src/setup/filesystem.sh:
```diff
@@ -0,0 +1,101 @@
+	if [ ! -d "$VAR_RUN" ] || [ ! -d "$VAR_TMP" ] || [ ! -d "$VAR_LOG" ]; then
+		echo "Configuring writeable partition..."
+		mkdir -p $VAR_RUN
+		mkdir -p $VAR_TMP
```
For sockets and tmp files

oparoz commented on src/setup/filesystem.sh:
```diff
@@ -0,0 +1,101 @@
+		echo "Configuring writeable partition..."
+		mkdir -p $VAR_RUN
+		mkdir -p $VAR_TMP
+		mkdir -p $VAR_LOG
```
For logs

oparoz commented on src/setup/filesystem.sh:
```diff
@@ -0,0 +1,101 @@
+	# Setup writable partition
+	if [ ! -d "$VAR_RUN" ] || [ ! -d "$VAR_TMP" ] || [ ! -d "$VAR_LOG" ]; then
+		echo "Configuring writeable partition..."
+		mkdir -p $VAR_RUN
```
For pids

oparoz commented on src/php/userconfigs/custom.conf:
```diff
@@ -0,0 +1,147 @@
+; For php_*flag, valid values are on, off, 1, 0, true, false, yes or no.
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+; ownCloud defaults
```
ownCloud default are modified here by the user

oparoz commented on src/php/php-fpm.conf:
```diff
@@ -0,0 +1,61 @@
+env[TMP] = ${SNAP_DATA}/var/tmp/
+env[TMPDIR] = ${SNAP_DATA}/var/tmp/
+env[TEMP] = ${SNAP_DATA}/var/tmp/
+php_admin_value[upload_tmp_dir] = ${SNAP_DATA}/var/tmp/
```
All the variables touching /tmp are set here

oparoz commented on src/php/php-fpm.conf:
```diff
@@ -0,0 +1,61 @@
+
+;;;;;;;;;;;;;;;;;;;;
+; Pool Definitions ;
+;;;;;;;;;;;;;;;;;;;;
```
This file is read-only

oparoz commented on src/php/php-fpm.conf:
```diff
@@ -0,0 +1,61 @@
+; Pass environment variables like LD_LIBRARY_PATH. All $VARIABLEs are taken from
+; the current environment.
+; Default Value: clean env
+clear_env = no
```
This needs to be validate by the security team and tested without.

oparoz commented on src/php/php-fpm:
```diff
@@ -0,0 +1,167 @@
+PHP_INI_SCAN_DIR=$RO_PHP_FOLDER/lib/php.conf.d:$PHP_USER_CONF
+export PHP_INI_SCAN_DIR
+
+OPTS="-R --daemonize --pid=${PID} --fpm-config $PHP_FPM_CONF -c $PHP_CONF -p $SNAP_DATA"
```
- Allow the process to be run as root
- Give location of main php.ini and php-fpm.conf
- Modify prefix

oparoz commented on src/php/php-fpm:
```diff
@@ -0,0 +1,167 @@
+
+# Tells PHP where to find its configuration files
+PHP_INI_SCAN_DIR=$RO_PHP_FOLDER/lib/php.conf.d:$PHP_USER_CONF
+export PHP_INI_SCAN_DIR
```
Adds additional config locations to PHP-FPM

oparoz commented on snapcraft.yaml:
```diff
@@ -55,53 +75,93 @@ parts:
     snap:
       - -manual # No need to include the documentation in the .snap
       - -htdocs/.git*
+      - -htdocs/.user.ini
```
Remove useless read-only .user.ini

kyrofa committed:
Upgrade to ownCloud 9.0.1. (#62)

Also started using the .htaccess file shipped in the root of ownCloud.
Note that this was done with a direct include rather than enabling
overrides due to the performance penalty. Typically this would be
unsafe, but the dangers are negated by the fact that the .htaccess file
is read-only.

This fixes #54.

Signed-off-by: Kyle Fazzari <kyle@canonical.com>

oparoz committed:
Use PHP-FPM instead of mod_php
Fixes #39
